### PR TITLE
First cut of migrating trace analytics plugins to event model

### LINKED ISSUE #612: OTel Group Prepper to Event
**Is your feature request related to a problem? Please describe.**

The Otel Group Prepper uses the deprecated string model.

**Describe the solution you'd like**

Update `otel-trace-group-prepper` to the `Event` model.

### LINKED ISSUE #613: OTel Raw Prepper to Event model
**Is your feature request related to a problem? Please describe.**

The OTel Raw Prepper uses the deprecated string model.

**Describe the solution you'd like**

Update `otel-trace-raw-prepper` to the `Event` model.

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
This PR is unavoidably lengthy for review but the main changes are outlined as follows:
* adapted Span Event model to facilitate migrating trace-analytics plugins, i.e. JacksonSpan, DefaultTraceGroupFields
* migrated the following plugins to span event model
* - otel-trace-source: previously only responsible for write ExportTraceServiceRequest(ETSR) into the Buffer. Now it is responsible for parsing the ETSR into collection of spans (OtelProtoCodec) and then offering them as buffer items
* - otel-trace-raw-prepper: previously responsible for parse ETSR (OtelProtoHelper) and do stateful processing on populating the traceGroup, traceGroupFields into all spans. Now only responsible for the latter.
* - otel-trace-group-prepper: responsibility remains the same -- making attempt to fill in missing traceGroup and traceGroupFields by looking up the opensearch backend
* modified the raw span e2e test and temporarily muted the e2e tests that involves service-map and peer-forwarder.

Note:
* In the second cut, OtelTraceSource might involve further changes depending on the client-side change, i.e. peer-forwarder.
* In the second cut, Span model might involve further API change or enrichment with the second cut of migration
* In the long term, we might consider merging otel-trace-raw-prepper and otel-trace-group-prepper as they share main responsibility now. But this is a breaking change not urgent right now.
 
### Issues Resolved
Contributes to #611 
Resolves #612 
Resolves #613 
Contributes to #546 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java [hunk 1]
```diff
@@ -220,36 +193,65 @@ public void testPrepareForShutdown() throws Exception {
         assertTrue(oTelTraceRawPrepper.isReadyForShutdown());
     }
 
-    private ExportTraceServiceRequest buildExportTraceServiceRequestFromJsonFile(String requestJsonFileName) throws IOException {
-        final StringBuilder jsonBuilder = new StringBuilder();
+    private static Span buildSpanFromJsonFile(final String jsonFileName) {
```
Comment (cmanning09) [inline data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java:196]: We originally did this to easily support the ExportTraceServiceRequests. Do we need to build the spans from a JSON file too? Can we build the Span objects we want to use for testing?

File: data-prepper-plugins/otel-trace-source/README.md [hunk 2]
```diff
@@ -88,8 +88,18 @@ curl -k -H 'Content-Type: application/json; charset=utf-8'  -d '{"resourceSpans"
...
+- `badRequests`: measures total number of requests with invalid format processed by otel trace source plugin.
+- `requestTimeouts`: measures total number of requests that time out.
+- `requestsTooLarge`: measures total number of requests of which the number of spans in the content is larger than the buffer capacity.
+- `internalServerError`: measures total number of requests processed by otel trace source with custom exception type.
+
+### Timer
+- `requestProcessDuration`: measures latency of requests processed by otel trace source plugin in seconds.
+
+### Distribution Summary
+- `payloadSize`: measures the distribution of incoming requests payload sizes in bytes.
```
Comment (cmanning09) [inline data-prepper-plugins/otel-trace-source/README.md:102]: What prompted this change? This seems outside of the scope of this work. Does this accomplish another issue that is open?

**Review by cmanning09 — CHANGES_REQUESTED** [review]
Hey @chenqi0805 , thanks for starting this. This PR is a large and it looks like there is a little scope creep here. Let's try to keep the PRs tightly scoped in the future.

I have added some feedback and comments. Given the size of this PR I think we need to get a second pair of eyes on this as well

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/trace/JacksonSpan.java [hunk 3]
```diff
@@ -43,10 +47,10 @@
     private static final String DURATION_IN_NANOS_KEY = "durationInNanos";
     private static final String TRACE_GROUP_FIELDS_KEY = "traceGroupFields";
 
+    private static final List<String> REQUIRED_KEYS = Arrays.asList(TRACE_GROUP_KEY);
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/trace/JacksonSpan.java:50]: The issue is traceGroup value could be null but we need the key as the placeholder.

File: data-prepper-api/src/test/java/com/amazon/dataprepper/model/trace/JacksonSpanTest.java [hunk 4]
```diff
@@ -385,6 +387,35 @@ public void testBuilder_withEmptyEndTime_throwsIllegalArgumentException() {
...
+                .withStartTime(TEST_START_TIME)
+                .withEndTime(TEST_END_TIME)
+                .withAttributes(TEST_ATTRIBUTES)
+                .withDroppedAttributesCount(TEST_DROPPED_ATTRIBUTES_COUNT)
+                .withEvents(Arrays.asList(defaultSpanEvent))
+                .withDroppedEventsCount(TEST_DROPPED_EVENTS_COUNT)
+                .withLinks(Arrays.asList(defaultLink))
+                .withDroppedLinksCount(TEST_DROPPED_LINKS_COUNT)
+                .withDurationInNanos(TEST_DURATION_IN_NANOS)
+                .withTraceGroupFields(defaultTraceGroupFields);
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-api/src/test/java/com/amazon/dataprepper/model/trace/JacksonSpanTest.java:409]: Here the test logic is about missing TRACE_GROUP key. `builder.withTraceGroup(null);` will still make the key present.

File: data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/model/TraceGroup.java [hunk 5]
```diff
@@ -10,54 +11,58 @@
     public static final String TRACE_GROUP_STATUS_CODE_FIELD = "traceGroupFields.statusCode";
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/model/TraceGroup.java:11]: Not following here. These constants are for look up the relevant doc field in opensearch. Could you specify the reasoning of their elimination?

File: data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java (hunk 1 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java:196]: Yes, it is just a test utility method. The intention is to keep the test data in files as much as possible instead of hardcode all the test values inline.

File: data-prepper-plugins/otel-trace-source/README.md (hunk 2 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/README.md:102]: Agree. This is kind of overreach. The initial intention is to add this metric for better monitoring. But to stick with the scope of this work I can pull it off.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/trace/JacksonSpan.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/trace/JacksonSpan.java:50]: Or are you suggesting we use empty string as placeholder? The only concern I have on that is if it is going to affect the trace-analytics UI display things.

File: data-prepper-plugins/otel-trace-source/README.md (hunk 2 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/README.md:102]: I might have misread. Do you mean all those new metrics? Some of them like `successRequests, badRequests` are migrated, transformed from processor:

```
### Counter
- `spanProcessingErrors`: records the number of processing exceptions for invalid spans.
- `resourceSpansProcessingErrors`: records the number of processing exceptions for invalid resource spans.
- `totalProcessingErrors`: records the total number of processing errors for spans and resource spans.
```

since we are now decoding the otel-proto request in source. While others like `requestsTooLarge`, `internalServerError` are due to the business logic shift caused by using Span model as Buffer item and calling `Buffer::writeAll` API. `requestProcessDuration` and `payloadSize` might be fully out of the scope although they are for better monitoring. Let me know what you think and I can pull ones off accordingly.

File: data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/model/TraceGroup.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/model/TraceGroup.java:11]: I was referring to eliminate the fields. I missed that these are used for look up. Thanks!

File: data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodecTest.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
+/*
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  The OpenSearch Contributors require contributions made to
+ *  this file be licensed under the Apache-2.0 license or a
+ *  compatible open source license.
+ *
+ *  Modifications Copyright OpenSearch Contributors. See
```
Comment (cmanning09) [inline data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodecTest.java:8]: This header doesn't seem correct given this is a brand new file.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
+package com.amazon.dataprepper.plugins.source.oteltrace.codec;
```
Comment (cmanning09) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:1]: Please include a copyright license here.

**Review by cmanning09 — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java (hunk 1 above)
Comment (cmanning09) [reply data-prepper-plugins/otel-trace-raw-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepperTest.java:196]: Thanks for the clarification. ExportTraceServiceRequests were larger objects when compared to our internal model. Building the test object from a JSON file made sense because the ETSR objects were so large. I am fine with keeping the files as long as we understand what we are optimizing for and we still consider our Spans to be large enough as well.

**Review by cmanning09 — APPROVED** [review]

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/event/JacksonEvent.java [hunk 8]
```diff
@@ -57,7 +57,7 @@
 
     private final EventMetadata eventMetadata;
 
-    private final JsonNode jsonNode;
+    protected final JsonNode jsonNode;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/event/JacksonEvent.java:60]: Once you make this protected, it is an exposed interface. Even for protected access, I recommend using methods rather than fields.

So keep this private and add `protected JsonNode getJsonNode()`.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java [hunk 9]
```diff
@@ -60,16 +90,45 @@ public void export(ExportTraceServiceRequest request, StreamObserver<ExportTrace
...
+            successRequestsCounter.increment();
             responseObserver.onNext(ExportTraceServiceResponse.newBuilder().build());
             responseObserver.onCompleted();
-        } catch (TimeoutException e) {
-            LOG.error("Buffer is full, unable to write");
-            requestTimeoutCounter.increment();
-            responseObserver
-                    .onError(Status.RESOURCE_EXHAUSTED.withDescription("Buffer is full, request timed out.")
-                            .asException());
+        } catch (Exception e) {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java:114]: I'd rather see specific catch statements for each Exception rather than an if-else in the exception handling. I'd prefer that even if the `LOG.error` is copied across all.

But, I wouldn't reject the PR over that.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    public static final Function<String, String> REPLACE_DOT_WITH_AT = i -> i.replace(DOT, AT);
+    /**
+     * Span and Resource attributes are essential for OpenSearch so they should not be nested. SO we will prefix them with "span.attributes"
+     * and "resource.attributes".
+     *
+     */
+    public static final Function<String, String> SPAN_ATTRIBUTES_REPLACE_DOT_WITH_AT = i -> SPAN_ATTRIBUTES + DOT + i.replace(DOT, AT);
+    public static final Function<String, String> RESOURCE_ATTRIBUTES_REPLACE_DOT_WITH_AT = i -> RESOURCE_ATTRIBUTES + DOT + i.replace(DOT, AT);
+
+    public static List<Span> parseExportTraceServiceRequest(final ExportTraceServiceRequest exportTraceServiceRequest) {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:65]: I recommend making this non-static. It is easier to test and extend the behavior of non-static methods.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        return exportTraceServiceRequest.getResourceSpansList().stream()
+                .flatMap(rs -> parseResourceSpans(rs).stream()).collect(Collectors.toList());
+    }
+
+    public static List<Span> parseResourceSpans(final ResourceSpans resourceSpans) {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:70]: Do all of these methods need to be public? It seems most or all of them can be private (or package protected if the tests need access).

Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:70]: Regarding my comment on non-static, I'm fine with keeping any private methods static.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java (hunk 9 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java:114]: This generic Exception has been tightly coupled to the exception thrown by writeAll API in the buffer. I will open a separate issue to track the enhancement from throwing generic exception to specific exceptions in buffer.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java (hunk 11 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:70]: I will make it protected. 

Just a note: OTelProtoCodec will be further refactored into common plugins in the following PR as for the convenience of maintenance on consistency between encoding (peer-forwarder client) and decoding (otel-trace-source). So this will not be the final version of it.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java (hunk 9 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java:117]: #882

File: data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcServiceTest.java [hunk 12]
```diff
@@ -72,38 +110,138 @@ public void setup() {
...
+        verify(responseObserver, times(0)).onCompleted();
+        verify(responseObserver, times(1)).onError(statusExceptionArgumentCaptor.capture());
+        verify(requestsReceivedCounter, times(1)).increment();
+        verify(badRequestsCounter, times(1)).increment();
+        verifyNoInteractions(successRequestsCounter);
+        verifyNoInteractions(requestsTooLargeCounter);
+        verifyNoInteractions(timeoutCounter);
+        final ArgumentCaptor<Double> payloadLengthCaptor = ArgumentCaptor.forClass(Double.class);
+        verify(payloadSizeSummary, times(1)).record(payloadLengthCaptor.capture());
+        assertEquals(SUCCESS_REQUEST.getSerializedSize(), Math.round(payloadLengthCaptor.getValue()));
```
Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcServiceTest.java:193]: Please use Hamcrest for new assertions.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+    /**
+     * Span and Resource attributes are essential for OpenSearch so they should not be nested. SO we will prefix them with "span.attributes"
+     * and "resource.attributes".
+     *
+     */
+    public static final Function<String, String> SPAN_ATTRIBUTES_REPLACE_DOT_WITH_AT = i -> SPAN_ATTRIBUTES + DOT + i.replace(DOT, AT);
+    public static final Function<String, String> RESOURCE_ATTRIBUTES_REPLACE_DOT_WITH_AT = i -> RESOURCE_ATTRIBUTES + DOT + i.replace(DOT, AT);
+
+    public List<Span> parseExportTraceServiceRequest(final ExportTraceServiceRequest exportTraceServiceRequest) {
+        return exportTraceServiceRequest.getResourceSpansList().stream()
```
Comment (dlvenable) [inline data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/codec/OTelProtoCodec.java:66]: Nice!

**Review by dlvenable — APPROVED** [review]
Looks good. Thanks for making changes per my previous requests.

File: data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java (hunk 9 above)
Comment (dlvenable) [reply data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcService.java:117]: My comment was more on the use of `if(e instanceof SpecificException)`. I'd rather: `catch(SpecificException e)`.

File: data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcServiceTest.java (hunk 12 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-source/src/test/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceGrpcServiceTest.java]: changed to hamcrest
